lib/templates/joins.genie

```diff
@@ -0,0 +1,86 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of Genie
+//
+// Copyright 2021 The Board of Trustees of the Leland Stanford Junior University
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Silei Xu <silei@cs.stanford.edu>
+
+import { Ast } from 'thingtalk';
+
+import * as C from './ast_manip';
+import ThingpediaLoader from './load-thingpedia';
+
+pronoun_the_first : string = {
+    'the first';
+    'the former';
+}
+
+pronoun_the_second : string = {
+    'the second';
+    'the latter';
+}
+
+each_other : string = {
+    'each other';
+    'one another';
+}
+
+short_phrase_table : Ast.Expression = {
+    complete_table;
+    one_clean_filter_table;
+}
+
+self_join_table : Ast.Expression = {
+    ( 'paris of' table:short_phrase_table ('such that'|'that') pronoun_the_first ('is'|'are'|'was'|'were') ('the'|'a'|'') param:out_param_Any 'of' pronoun_the_second 
+    | 'paris of' table:short_phrase_table ('such that'|'that') pronoun_the_first ('has'|'have'|'had') pronoun_the_second 'as' ('the'|'a'|'') param:out_param_Any 
+    | table:short_phrase_table 'pairs' ('such that'|'that') pronoun_the_first ('is'|'are'|'was'|'were') ('the'|'a'|'') param:out_param_Any 'of' pronoun_the_second 
+    | table:short_phrase_table 'pairs' ('such that'|'that') pronoun_the_first ('has'|'have'|'had') pronoun_the_second 'as' ('the'|'a'|'') param:out_param_Any 
+    ) => C.makeSelfJoinFromParam($loader, table, param);
```

Comment:
So this generates (among other):
"male people pairs such that the first is the spouse of the second"
is that the expected sentence?
"X pairs" sounds awkward to me, especially with an adjective in front.

It also generates
"male people pairs that the first is the spouse of the second"
(using "that" instead of "such that").
I think this one is just wrong.

It also generates
"people born in X pairs such that the first is the spouse of the second"
(using the "pvp" form of one_clean_filter_table)
I think this one is also just wrong.

Reply:
yeah you are right I removed the whole "X pairs" thing from the templates